In [2]:
from sklearn import datasets, metrics
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split

In [3]:
# 讀取鳶尾花資料集
iris = datasets.load_iris()

# 切分訓練集/測試集
x_train, x_test, y_train, y_test = train_test_split(iris.data, iris.target, test_size=0.25, random_state=4)

# 建立模型
clf = RandomForestClassifier()

# 訓練模型
clf.fit(x_train, y_train)

# 預測測試集
y_pred = clf.predict(x_test)

In [4]:
acc = metrics.accuracy_score(y_test, y_pred)
print("Acuuracy: ", acc)

Acuuracy:  0.9473684210526315


In [5]:
print(iris.feature_names)

['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


In [6]:
print("Feature importance: ", clf.feature_importances_)

Feature importance:  [0.02850691 0.01305996 0.51549285 0.44294028]


# 作業   
 **1. 試著調整 RandomForestClassifier(...) 中的參數，並觀察是否會改變結果？**


In [9]:
# 建立模型
clf = RandomForestClassifier(
    n_estimators=50,
    criterion="gini",
    max_features="auto", 
    max_depth=10,
    min_samples_split=2,
    min_samples_leaf=1
                            )

# 訓練模型
clf.fit(x_train, y_train)

# 預測測試集
y_pred = clf.predict(x_test)

acc = metrics.accuracy_score(y_test, y_pred)
print("Acuuracy: ", acc)
print(iris.feature_names)
print("Feature importance: ", clf.feature_importances_)

Acuuracy:  0.9736842105263158
['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']
Feature importance:  [0.07572663 0.02358993 0.46647356 0.43420988]


→調整參數後，acuuracy從0.947上升到0.973。

 **2. 改用其他資料集 (boston, wine)，並與回歸模型與決策樹的結果進行比較**

In [9]:
Breast_cancer = datasets.load_breast_cancer()
wine = datasets.load_wine()

**Breast_cancer**

In [12]:
print(Breast_cancer.data.shape)
print(Breast_cancer.feature_names)

# 切分訓練集/測試集
x_train, x_test, y_train, y_test = train_test_split(Breast_cancer.data, Breast_cancer.target, test_size=0.25, random_state=4)

# 建立模型
clf = RandomForestClassifier()

# 訓練模型
clf.fit(x_train, y_train)

# 預測測試集
y_pred = clf.predict(x_test)

acc = metrics.accuracy_score(y_test, y_pred)
print("Accuracy: ", acc)

print(Breast_cancer.feature_names)
print("Feature importance: ", clf.feature_importances_)

(569, 30)
['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'worst concave points' 'worst symmetry' 'worst fractal dimension']
Accuracy:  0.9370629370629371
['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst 

→與之前比較：  
[Logistic](https://github.com/Erincatcat/ML100-Days/blob/master/Homework/Day_038_HW.ipynb) accuracy=0.965  
[Decision tree](https://github.com/Erincatcat/ML100-Days/blob/master/Homework/Day_042_HW.ipynb) accuracy=0.909  
Random forest accuracy=0.937  


**Wine**

In [13]:
print(wine.data.shape)
print(wine.feature_names)

# 切分訓練集/測試集
x_train, x_test, y_train, y_test = train_test_split(wine.data, wine.target, test_size=0.25, random_state=4)

# 建立模型
clf = RandomForestClassifier()

# 訓練模型
clf.fit(x_train, y_train)

# 預測測試集
y_pred = clf.predict(x_test)

acc = metrics.accuracy_score(y_test, y_pred)
print("Accuracy: ", acc)

print(wine.feature_names)
print("Feature importance: ", clf.feature_importances_)

(178, 13)
['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium', 'total_phenols', 'flavanoids', 'nonflavanoid_phenols', 'proanthocyanins', 'color_intensity', 'hue', 'od280/od315_of_diluted_wines', 'proline']
Accuracy:  0.9555555555555556
['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium', 'total_phenols', 'flavanoids', 'nonflavanoid_phenols', 'proanthocyanins', 'color_intensity', 'hue', 'od280/od315_of_diluted_wines', 'proline']
Feature importance:  [0.01865953 0.01729882 0.01507511 0.03711935 0.01821504 0.00740665
 0.12383618 0.00393375 0.01013669 0.27053144 0.02200428 0.23311387
 0.2226693 ]


→與之前比較：  
[Logistic](https://github.com/Erincatcat/ML100-Days/blob/master/Homework/Day_038_HW.ipynb) accuracy=1    
[Decision tree](https://github.com/Erincatcat/ML100-Days/blob/master/Homework/Day_042_HW.ipynb) accuracy=0.911    
Random forest accuracy=0.955  


兩份資料的accuracy都是Logistic表現最好，Random forest次之，Decision tree最末。

課後補充:

[知名 ML youtuber 教你手刻隨機森林 by Python](https://www.youtube.com/watch?v=QHOazyP-YlM#action=share)